Movie extras

debmint opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://jellyfin.readthedocs.io/en/latest/developer-docs/contributing/ page.
-->
Add Extended Details to Movies and TV Shows

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->
Add Cast & Crew, "More Like This", and "Special Features" to Movies and TV Shows.  "Special Features" includes any extra videos, such as "Behind the Scenes", etc. that might be in the library.

**Issues*
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->
This adds a feature that is missing in the app

neilsb commented:
This is looking really good.  Thanks.  I'll try and give it a proper test and review this week at some point.

debmint commented:
Good to hear.  It may need some fixing.  As I understand it, the PR was to pull to master, but if it is mostly acceptable,it might be good to pull to its own branch and let others make improvements.

neilsb commented:
Sorry for the delay in getting this looked at.  I think it works well, and wouldn't have a problem rolling it into main client and let it be worked on there, rather than having a separate branch.  Picked up a couple of things while testing, but I don't think there's any risk that this is going to cause any problems making the client unusable, so don't see any reason not to include it in the main version.

A few minor things
 - On the Movie Details screen, the Cast & Crew is visible at the bottom faded out, indicating to the user that there is something to navigate to.  Is there any reason this isn't also the case on the Actor Page?  Folk might not notice that additional panel is available there.
 - From the Actor Screen, the TV Shows use the episode thumbnail icon, which is widescreen, but it's resized into portrait.   
 - From the Actor Screen, selecting a TV Show gives a "This type is not supported" message.   Also, when you clear the message focus is lost and you need to use back

The linter is not happy with the code formatting (still not sure if we want to turn this off) - Can run `npm run check-formatting` to see what it's not happy about.  The easiest way to sort this is to (ensuring you've checked everything in first) is to run `npm run format`, and it'll make any formatting changes and you can give it a once over before committing.

I'll get a proper review of the code done shortly, but this looks really good.

debmint commented:
> Sorry for the delay in getting this looked at. I think it works well, and wouldn't have a problem rolling it into main client and let it be worked on there, rather than having a separate branch. 
That's good to hear.  I would like to see these mods I've made be incorporated if they are not exceedingly bad coding.
>Picked up a couple of things while testing, but I don't think there's any risk that this is going to cause any problems making the client unusable, so don't see any reason not to include it in the main version.
>
> A few minor things
> 
>     * On the Movie Details screen, the Cast & Crew is visible at the bottom faded out, indicating to the user that there is something to navigate to.  Is there any reason this isn't also the case on the Actor Page?  Folk might not notice that additional panel is available there.
I had thought about this.  I call this group "Extras" and most of this is defined in components/extras", and the slider is "ExtrasSlider".  In all the screens I place the extras slider below the outer LayoutGroup for all the other stuff.  It appeared I had to do this to get the slider to work.  I'm afraid my understanding of ScreenGraph is still quite limited.  I gave the slider a y-translation of 960, and I gave the slider a bit of visibility in order to let the user see that there _is_ something there.  In the PersonInfo, for some reason, this is not working.  I will investigate.  I hadn't made any changes since posting PR to avoid any possible confusion to reviewers.  I did rm that one file yesterday to remove one error reported by the test.

>     * From the Actor Screen, the TV Shows use the episode thumbnail icon, which is widescreen, but it's resized into portrait.
Should these be left at a wider proportion? 
>     * From the Actor Screen, selecting a TV Show gives a "This type is not supported" message.   Also, when you clear the message focus is lost and you need to use back
My bad here, I guess.  Must not have even checked that.  I may have to change the "Type".  The popup done by code that's already there.  I set my stuff to use the already-existing callbacks, and these are all in the Main.brs file.
> The linter is not happy with the code formatting (still not sure if we want to turn this off) - Can run `npm run check-formatting` to see what it's not happy about.
I saw those.  Most seem to be simple pedantic coding style.  I had several "then" statements at the end of "if" statements.  I saw in some diffs where someone did add "then"s to several statements so assumed that this was preferred.  This could be easily remedied.  I saw one place where it complained about  too many parentheses in an if statement.  Again, easily fixed if it's a problem..  May be other things, too.  I did remove the one unused file that contained what it deemed an  uninitialized var.
> The easiest way to sort this is to (ensuring you've checked everything in first) is to run `npm run format`, and it'll make any formatting changes and you can give it a once over before committing.
Ok, will do that.  I'm fine with doing all the discussion here, but if it's too much for here, feel free to contact me via my gmail account under debmint49.
> I'll get a proper review of the code done shortly, but this looks really good.

neilsb commented:
> I had thought about this. I call this group "Extras" and most of this is defined in components/extras", and the slider is 

The change you've made looks good. Lets the user know there's something more on the actor page now.

> > * From the Actor Screen, the TV Shows use the episode thumbnail icon, which is widescreen, but it's resized into portrait.
> 
> Should these be left at a wider proportion?

Ideally yes, as that's generally how episode images are displayed.  However, if it's problematic then the TV Show itself should have a portrait poster that can be used, instead of an episode specific file.

> > * From the Actor Screen, selecting a TV Show gives a "This type is not supported" message.   Also, when you clear the message focus is lost and you need to use back
> 
> My bad here, I guess. Must not have even checked that. I may have to change the "Type". The popup done by code that's already there. I set my stuff to use the already-existing callbacks, and these are all in the Main.brs file.

That's working fine now. Thanks  👍

debmint commented:
> > > * From the Actor Screen, the TV Shows use the episode thumbnail icon, which is widescreen, but it's resized into portrait.
> > 
> > Should these be left at a wider proportion?
> 
> Ideally yes, as that's generally how episode images are displayed. However, if it's problematic then the TV Show itself should have a portrait poster that can be used, instead of an episode specific file.

I'll change the proportions for these.

neilsb commented:
> I'll change the proportions for these. (Can't do it just now - perhaps sometime tonight)

That's looking good now.  👍 	

> Another thing, should I put a limit on the number of shows/movies to fetch? The list for a long-running series would be quite long. If so, what would be a reasonable count?

Yeah, that makes.  Since it's not a screen that will be used all the time, I don't think it has to be a very small number to keep it snappy.  Maybe 50 - 100 or something like that?  Can always be adjusted based on feedback from users if required.

Functionally it's all looking good now.  I'll try to though the code this weekend.  Thanks again.

jimdogx commented:
@debmint I just tested this and this is great!!!

There is one issue though.  On the cast information screen, if you click on Options it crashes.  If there are no options here, might be easiest to simply hide the option.

The only other thing is if the cast member has a rather lengthy information page, there is no way to "Show More" like on the web client.  This to me is very minor though.

Here is the screen I'm talking about (both the crash and "Show More"):

![Screen Shot 2022-02-19 at 6 34 04 AM](https://user-images.githubusercontent.com/18198517/154803192-dc3ec922-5e7e-4980-93a7-5dca3d19329e.png)

debmint commented:
> @debmint I just tested this and this is great!!!

The screenshots are a little larger than those on the GridView, for example, but it seems that I have set the proportions close to what they should be.

> There is one issue though. On the cast information screen, if you click on Options it crashes. If there are no options here, might be easiest to simply hide the option.

I need to get rid of the Options menu.  The screen is based on JFGroup and it has a field "optionsAvailable".  Maybe setting this to false might do this. 

> The only other thing is if the cast member has a rather lengthy information page, there is no way to "Show More" like on the web client. This to me is very minor though.

I had noticed this.  I will have to figure out how to go about this.  I think setting the height of the label to a lesser value will shorten it, but then will need to handle showing the full description.  I had thought about a "Show More" button that would pop up a scrollable dialog.  The trouble with trying to mimic the web client is that a web page automatically scrolls where, if I understand correctly, with the Roku client, a screen taller than the display simply hides the excess data.  That's why I had to go the route I did in enabling the Extras.  I will see what I can do.

jimdogx commented:
For the "Show More" issue maybe this will work:

https://developer.roku.com/docs/references/scenegraph/typographic-nodes/scrollabletext.md

I've never used it before, so no guarantees 😄

debmint commented:
> For the "Show More" issue maybe this will work:

> I've never used it before, so no guarantees smile

> Edit: I tested ScrollableText real quick and it does work. However, it's a bit more than just replacing the description "Label" with "ScrollableText". You'll also have to modify onKeyEvent() in PersonsInfo.brs since right now "up" and "down" show/hide the other panel and doesn't let you control the scrollbar.

Believe it or not, that's just what I've been working on.  I haven't pushed it yet, or even committed it on my local.  I can  commit and push it but thought it might clutter up the history if I had to revert, although that would be OK with me.  Let me see if I can go over what I've done and see if you want me to push it.  I can transfer these changes over to a separate  branch, too if keeping the history clean is a priority.  Anyway...

I have disabled options - simple as setting "optionsAvailable" to false.
I then set the description to a ScrollableText which is scrollable if the text is truncated - otherwise it's just plain old text.  I then changed the "onKeyEvent" to check for the "options" keypress.  At the top right of the screen, under the time, I set a label text to "* " and label "showVidText", with default text "Show Videos".  In "onKeyEvent" the popup is the same except it checks the text of "showVidText".  If the slider is being displayed, the text is toggled to "Hide Videos".

I fired up Emby and in this scene it displays the description as truncated if it is long.  The focus is on "Favorites"  Pressing "UP" sets focus on the description text.  Pressing "OK" pops up a dialog of some type with ScrollableText.  "BACK" gets you out of that.  This seems like a pretty decent setup - I had considered something similar.

Does the way I did it sound reasonable?  Do you want me to push what I did?  Should I send it up in a separate branch for evaluation before adding it to movieExtras?  My biggest concern is that having to use "*" here might be confusing to the user.

jimdogx commented:
> I have disabled options - simple as setting "optionsAvailable" to false.

👍 

> I then set the description to a ScrollableText which is scrollable if the text is truncated - otherwise it's just plain old text. I then changed the "onKeyEvent" to check for the "options" keypress. At the top right of the screen, under the time, I set a label text to "* " and label "showVidText", with default text "Show Videos". In "onKeyEvent" the popup is the same except it checks the text of "showVidText". If the slider is being displayed, the text is toggled to "Hide Videos".

To be honest, I was actually thinking of something way simpler (but maybe too simple?)  Something along the lines of:

(pseudo code)
if key = "right" hide videos // control scrollbar from here on out
if key = "left" show videos // up and down now show/hide extra info pane

Your way might actually be more obvious to a user since it has a label.

> Do you want me to push what I did? Should I send it up in a separate branch for evaluation before adding it to movieExtras? 

I don't know if clean history is a priority.  I've had a PR or two that have had several "corrections". 😄   But it's really someone else's call (Neil etc).

jimdogx commented:
> I have disabled options - simple as setting "optionsAvailable" to false.

👍 

> I then set the description to a ScrollableText which is scrollable if the text is truncated - otherwise it's just plain old text. I then changed the "onKeyEvent" to check for the "options" keypress. At the top right of the screen, under the time, I set a label text to "* " and label "showVidText", with default text "Show Videos". In "onKeyEvent" the popup is the same except it checks the text of "showVidText". If the slider is being displayed, the text is toggled to "Hide Videos".

To be honest, I was actually thinking of something way simpler (but maybe too simple?)  Something along the lines of:

(pseudo code)
if key = "right" hide videos // control scrollbar from here on out
if key = "left" show videos // up and down now show/hide extra info pane

Your way might actually be more obvious to a user since it has a label.

> Do you want me to push what I did? Should I send it up in a separate branch for evaluation before adding it to movieExtras? 

I don't know if clean history is a priority.  I've had a PR or two that have had several "corrections". 😄   But it's really someone else's call (Neil etc).

jimdogx commented:
> I don't know if clean history is a priority. I've had a PR or two that have had several "corrections".

I should have done a better job reading the developer guidelines.  A "clean history" is actually desired.  FYI:

https://jellyfin.org/docs/general/contributing/development.html#pull-request-guidelines

debmint commented:
> > I don't know if clean history is a priority. I've had a PR or two that have had several "corrections".
> 
> I should have done a better job reading the developer guidelines. A "clean history" is actually desired. FYI:

From the guidelines, it seems like they want as little in the history as possible, but that comes from the main Jellyfin repo.  They might not be as strict here, I don't know.  Maybe it would be best to get Neil's input.  I've already accumulated several commits in this branch, although I've always thought that you don't want to jumble a whole bunch of perhaps unrelated changes in one commit, and the guideline even sort of agrees with that.  I just wonder if I'm going in a direction that the eaders might not want to go.   I could stash these changes, git stash branch a temporary branch then restash this state again because this state would be popped by the branch.  That way I could resume the changes in this branch if it was accepted.  I would, of course push the new branch for you all to test, and drop that branch after everyone saw it.  This way, if this route was a no-go, it wouldn't be in the history.

The only thing that bothers me about what I have now is that in the other places, to get the Extras up is with the DOWN arrow, but here, it's with the STAR.  I find myself wanting to do the DOWN arrow here, too.

neilsb commented:
> From the guidelines, it seems like they want as little in the history as possible, but that comes from the main Jellyfin repo. They might not be as strict here, I don't know. Maybe it would be best to get Neil's input.

I wouldn't worry about it too much.  The main Jellyfin repo has a significantly different number of people committing to it and having to keep up to date with it.

Additionally, I think that commits resulting in feedback from a PR are fine - think the guidelines are primarily aimed at the number of commits made prior to the PR submission.

And in the end, we can always do a Squash merge if it's felt there are too many individual commits.

neilsb reviewed (changes requested):
Code mostly looks fine, just a few things to update to ensure the UI is translated into the appropriate language for the users current locale.

Additionally, there's a significant update to the package-lock.json file (updating format from v1 to v2).  This probably is probably something we should do, so can be left in here for now.  But just for future reference if changes aren't directly related to the submitted PR, they should probably be in another PR of their own.

neilsb commented on components/home/LoadItemsTask.brs:
```diff
@@ -100,8 +100,83 @@ sub loadItems()
+
+        ' Extract array of persons from Views and download full metadata for each
+    else if m.top.itemsToLoad = "people"
+        for each person in m.top.peopleList
```
This section doesn't appear to perform any API call.  The `LoadItems` task was initially created to prevent the UI from locking up while a server API call was being made.  

Not sure if the overhead to create and run a task is beneficial here, since it's a for loop on local data.  However, if it beneficial perhaps it would be better to create a new Task for this data manipulation, and keep the `LoadItems` task specifically for loading things from the server?

debmint replied:
> This section doesn't appear to perform any API call. The `LoadItems` task was initially created to prevent the UI from locking up while a server API call was being made.
> 
> Not sure if the overhead to create and run a task is beneficial here, since it's a for loop on local data. However, if it beneficial perhaps it would be better to create a new Task for this data manipulation, and keep the `LoadItems` task specifically for loading things from the server?

I don't know if it's still applicable to the current build I have, but early on in my attempts, when I just called up the list of persons, before I implemented the callback, just having the row of cast and crew, if I repeatedly clicked on movies, after a few clicks, I would get an error where, in components/api/baseReqest.brs/buildParams()/ where "req" (a roUrlTransfer) did not get created (= invalid).  imageURL()->buildURL()->buildParams().  This may not happen now that I have it up and running.

jimdogx replied:
I hadn't paid much attention to this comment by Neil (or your update) until now.

I think Neil missed the ImageUrl() call.  That indeed is an API call.  So it does need to be here in LoadItemTask 👍

debmint replied:
Yes, I really don't know what's going on here.  I've tried to trace out the existing calls.  A lot of the calls to ImageURL() come in the modules.  A lot of the modules are called from source/ShowScenes.brs, which is called from Main.brs.  These call ItemMetadata() which sets a field "image" that contains the uri, which would satisfy the MAIN/TASK requirement.  The funny thing, I can often display a Detail a couple or so times before it crashes.  I've tried it on what I now have pushed to GitHub and it doesn't seem to have any problems, but I have everything AFAIK done inside TASKS

I do believe I have some things that should be streamlined, however.  I'm wondering if I should have started my own task file.. LoadItemTasks is really designed for the HOME group.  I was just trying to avoid adding any more files than necessary,  in examining the code base, I've wondered if a lot of the files could be generalized and used in multiple instance.  It might be complicated, though.. I initially tried to implement some existing files, but found it would take quite a bit of modification.  I didn't want to change any existing stuff on my own (although I did make a few changes that I felt I had to do).

neilsb replied:
Yeah, sorry.  Totally missed that ImageUrl call was an API call.  Sorry about that.

I personally don't have problem with leaving it in the LoadItems task right now if that's easier.  Things can always be refactored later on.   If it's being left in the LoadItems task, can I suggest that the `peopleList` property is named something more generic, so it _could_ be used for similar tasks that are not people.  Maybe something like `itemData`?

debmint replied:
Renaming the var to "itemData" sounds fine, although I'm now wonder if it would be best to move all the stuff I've added out into its own file - probably in the "extras" directory.  My first intent was to not add a bunch of files, but after seeing the system, it might be more in line with the structure to keep it segregated.

I aded one function in the TASK that simply gets the ImageURL, only called fromPersonInfo.  This object is created in a function in source/ShowScene.brs.  I wonder if the image url could be set up from there, either directly or maybe calling ItemMetadata or something.  I had added a function in ItemMetadata but haven't called it yet.  If the API functions are called from here, (which is in the MAIN thread (?), it will work - right?  I'm still trying to figure out how to avoid calling the "people" TASK.

@jimdogx and I have been discussing the way I implemented the person description.  Do you have any thoughts on that?

debmint commented:
> Additionally, there's a significant update to the package-lock.json file

This must have come in when I ran some npm command.  So it's no problem if I leave this in?

debmint commented on components/PersonInfo.brs:
```diff
@@ -0,0 +1,97 @@
+        birthDate = CreateObject("roDateTime")
+        birthDate.FromISO8601String(data.PremiereDate)
+        deathDate = CreateObject("roDatetime")
+        lifeString = "Born: " + birthDate.AsDateString("short-month-no-weekday")
```
I think I've gotten all my strings under tr()

> Will also need to add the string to the `locale\en_US\translations.ts` file, which is the master file and will make the string available on Weblate automatically for translation

Do I manually add these to the file? If so, does it matter where I put them?  Top? Bottom?   Do I add <message> and <source>? Do I add the <translation> or does the WebLate handle this?

neilsb replied:
Doesn't matter where in the file it goes. Generally just add new stuff to the bottom. 

Just need to add a complete entry in the en_US file (the description bit is not required, but may be helpful to give context to the folk that translate)

Weblate will then pick it up and allow users to translate it for any other languages that are set up on the project.

debmint commented:
I have pushed the changes discussed above.  Check out the new layout for Persons.  I'm not convinced this is the best route.  To pop up the Movies/Sthe hows, the STAR or something besides DOWN needs to be pressed because if ScrollingText is in effect, ST ties up the UP/DOWN buttons.

The alternative is to display description as truncated if it is long.  When description has focus click OK to display a dialog inscrollable text mode.  This would free up DOWN key to display RowList.  The way Emby does it is initially there is a "Favorites toggle button that has focus.  UP gives focus to description, DOWN brings up RowList.  However, would anyone have interest in having Favorites for persons?  I believe I can do all this but is it worth the effort?

jimdogx commented:
Just tried it.  It is a little different having to hit "*", but I like that the "*" is in a color that stands out so it drew my eye and was easy to understand once I used the feature.  I think our users are smart enough to figure it out 😄   Up to Neil though.

debmint commented:
> Just tried it. It is a little different having to hit "*", but I like that the "*" is in a color that stands out so it drew my eye and was easy to understand once I used the feature.
> 
> Wondering how many cast members have lengthy bios like "The Rock". Because if it's a rare thing, users might wonder why they have to hit "*". It makes sense when you're on a person that requires scrolling though.

I think the occurrence is not rare.

> I think our users are smart enough to figure it out smile Up to Neil though.

The prompt should make it clear, but still, it does depart from the "norm" so to speak.  Maybe we will get more input on this.  Repeating what I said before, I'm wondering if it would be worthwhile to include a "Favorites" feature for actors etc.

neilsb commented:
> > Additionally, there's a significant update to the package-lock.json file
> 
> This must have come in when I ran some npm command. So it's no problem if I leave this in?

Yeah, fine to leave in.

debmint commented:
> The Person Detail page with the scroller makes sense, and seems like a good way to implement this page at the current time.
> 
> Had another look over the code, and everything seems good. I'm happy for this to be merged now if everyone else is OK with it?

I have another commit that I haven't pushed yet.  Doesn't change anything visibly, but I've restructured the Person Info to be more like the rest of the jellyfin-roku code.  I've renamed PersonInfo to PersonDetails, and implemented a PersonData ContentNode, going through ItemMetadata().

Additionally, I am working on a different approach to the PersonDetails scene.  I have the details, or overview, initially displayed truncated.  Below this I have a "Favorites" Button where you can toggle whether this person is a favorite.  ATM, on entry, this button has focus.  Pressing UP sends focus to description, where, if truncated, opens a StandardMessageDialog filling most of the screen where you can scroll through the text.  I have this part mostly working but I don't have Favorites setting itself on the server yet.  I'm doing this in a separate branch.  If this sounds interesting I can push it in its own branch and if approved, can then merge it into movieExtras.  I'm sort of getting used to pressing STAR to navigate in the current setup but it's still feels a little odd to me.  As for Favorite for people, would this be something in which many would be interested?  I could easily remove it and set initial focus on the description and DOWN arrow to extras

debmint commented:
> Not a problem at all. Think the person favourites is a good feature to have, so happy for it to be included.
> 
> Don't mind how you want to do it, either just push it to this PR or create a new one.
> 
> I was just conscious I hadn't reviewed your PR and didn't want you to think it had been forgotten or wasn't valued. If you want to add more it now (either the refactoring or both that and the favourite) then we'll hold off the merge for now. Not a problem.

Well, this has given me more time to do some improving (I hope they are).

> If you're still working on stuff, can mark the PR as draft and then when you're ready for it it to be merged you can set it back as ready. That lets you control when it's merged blush

I hope I have it relatively complete for now.  At the moment, I have played around with Revision Control Systems for a long time, but this is my first attempt at collaborative work and my main concern is providing clean work.  Git has some pretty powerful features (some of which I am just now finding out about) but as i understand, it's a bit more complicated to modify after its pushed - or can cause problems anyway.

Would it be best to just push to a new  branch, let you all checkout the branch and if you like it, then merge or rebase it onto this branch, or if it's not acceptable, then I can just delete it.  Or, if you prefer, I can merge/rebase it here and I suppose we can revert it if not approved.  Which would be best?

I think I have Favorites working now, and I believe I have the scene working like I wanted.

jimdogx commented:
> Would it be best to just push to a new branch, let you all checkout the branch and if you like it, then merge or rebase it onto this branch, or if it's not acceptable, then I can just delete it. Or, if you prefer, I can merge/rebase it here and I suppose we can revert it if not approved. Which would be best?

My $0.02 is to merge what you have to this PR.  I'd love to checkout the latest changes and play around / test it like I did previously.  It sounds like the new changes will most likely get approved (barring any code cleanup etc).

And if for some reason it's really, really horrible (doubtful), you can revert the commit and force push back to the branch.  This would be a big no-no if more than one person was working on this feature branch, but since you're the only one, it should be fine.

debmint commented:
> My $0.02 is to merge what you have to this PR.

OK that's what I'll do.  I just pushed some changes I had made in the original branch.  I summed up those in a msg to Neil - basically changed structure for files.

jimdogx commented:
A couple of comments / questions:

1.  I'm wondering if the faded pane can go down just a tad bit more?  If you look at a show with multiple seasons, the faded pane overlaps with the "Season 1", "Season 2" text at the bottom.  I can provide a screenshot to explain if needed.

2. The actor details appear to only search for movies.  So if the actor is in both movies and TV shows, it only shows the movies.  And if the actor is only in TV shows then the faded pane is empty (I can also share a screenshot if needed).  You'll notice the Web Client shows both Movies and TV shows.

3. If #2 isn't possible or too much to tackle now, can the pane be smart enough to hide itself if it's empty?  Although in theory, if you fix #2 then the pane would never be empty (because at the very least it would show the movie and/or TV Show you're currently on).

debmint commented:
> A couple of comments / questions:
> 
>     1. I'm wondering if the faded pane can go down just a tad bit more?  If you look at a show with multiple seasons, the faded pane overlaps with the "Season 1", "Season 2" text at the bottom.  I can provide a screenshot to explain if needed.

It could.  However, on my screen (and probably all), the Season 1... goes practically all the way to the bottom.  I want to leave enough of the stuff to let users know there is something there.  we could lower it some, but I wonder if the other stuff could be shortened a bit, too
> 
>     2. The actor details appear to only search for movies.  So if the actor is in both movies and TV shows, it only shows the movies.  Do you have the latest pushes?  My library is very tiny, but I have some actors who appear in both and they show both

>     3. If item number 2 isn't possible or too much to tackle now, can the pane be smart enough to hide itself if it's empty?  Although in theory, if you fix item number 2 then the pane would never be empty (because at the very least it would show the movie and/or TV Show you're currently on).

Will look into this, but this is not the case with what I have

> p.s. sorry I didn't get these comments to you sooner. I knew you were still working on it so I wanted to wait until you had more stuff finished / polished before re-testing.

I will check it out, but do be sure you don't have an outdated version.  I didn't think I pushed anything that did as you describe

jimdogx commented:
> EDIT: I just pushed all my latest stuff. Had to do a pull and Had to edit several conflicts. Some of the commits were yours and I hope I didn't get the stuff backward. You might need to check what I did One example, There were some places where you either added or deleted some show and hide of overhang. I left them in. Hope that was right.

I'm not sure you did the merge correctly.  It's missing my last commit that cleaned up the overhang stuff.  When you did a pull, was your fork up-to-date with Master?

I'm used to just seeing one "Merged from Master" commit, instead I'm seeing my individual commits added to your PR (with some missing).

You might need to go back to [8eea8e4](https://github.com/jellyfin/jellyfin-roku/pull/520/commits/8eea8e4134971f1c4d737013613dbef77f26bce4) and then fetch upstream / merge and re-push.

debmint commented:
> I'm not sure you did the merge correctly. It's missing my last commit that cleaned up the overhang stuff. When you did a pull, was your fork up-to-date with Master?
> 
> I'm used to just seeing one "Merged from Master" commit, instead I'm seeing my individual commits added to your PR (with some missing).

I _thought_ Neil had merged master into this branch, but I might be wrong.  What did commit 53c23a0 (on my fork) do?

> You might need to go back to [8eea8e4](https://github.com/jellyfin/jellyfin-roku/pull/520/commits/8eea8e4134971f1c4d737013613dbef77f26bce4) and then fetch upstream / merge and re-push.

OK .. so what I need to do is reset --hard 8eea8e4?  Then checkout master?  Do "git pull upstream master" "checkout movieExtras" ... "merge master"? Then I'll have to "git push --force" won't I?  If all this isn't right, please step me through it

jimdogx commented:
> I thought Neil had merged master into this branch, but I might be wrong. What did commit https://github.com/jellyfin/jellyfin-roku/commit/53c23a0b53ef24864b615861fbf4f6376cfa0662 (on my fork) do?

Yes, it looks like Neil had already merged from Master, so in theory you shouldn't have had to do anything on your part (i.e. those two extra commits of mine that were added yesterday).  Not sure why you had conflicts 😦 

> OK .. so what I need to do is reset --hard https://github.com/jellyfin/jellyfin-roku/commit/8eea8e4134971f1c4d737013613dbef77f26bce4? I will need to do a hard reset, won't I? Then checkout master? Do "git pull upstream master" "checkout movieExtras" ... "merge master"? Then I'll have to "git push --force" won't I? If all this isn't right, please step me through it

> EDIT: I did the above and after doing pull, it went much smoother. After doing this, I cloned jellyfin/jellyfin-roku into another directory, did a diff on my movieExtras against upstream master and the only differences were my changes, except for one or two lines in your changes that looked identical to me. So now just a matter of "git push --force" ???

Sounds about right.

debmint commented:
> > I thought Neil had merged master into this branch

> Yes, it looks like Neil had already merged from Master, so in theory you shouldn't have had to do anything on your part

I did pull from upstream and it looked like all that was added was the change to README

> (i.e. those two extra commits of mine that were added yesterday). Not sure why you had conflicts frowning

I don't either.  For the most part, I did git rebase as that seemed to be what was recommended in the PR guidelines.  However, when I did it the last time, it went a lot smoother.

>  So now just a matter of "git push --force" ???
> Sounds about right.

I'll get on it.  On the part of you guys, if you don't have any changes in your working dirs for my fork - I assume you would have one, I'd suppose it would be best for you to just delete your copies and just clone again?  I'd say  pull would report divergent branch.  So here goes...

debmint commented:
> A couple of comments / questions:
> 
>     1. I'm wondering if the faded pane can go down just a tad bit more?  If you look at a show with multiple seasons, the faded pane overlaps with the "Season 1", "Season 2" text at the bottom.  I can provide a screenshot to explain if needed.

What do you suggest?  I could cut the top margin down a bit, but if I cut it down too much, the text at the top will look crowded.  I can drop the whole thing down but I want enough to be visible for the user to realize that something is there.  Neil mentioned this when I originally had it invisible until you clicked down arrow.  I could make it more transparent but I want it to be visible enough for the user to see it well enough to realize it is there.
The Seasons row is, itself a RowList.  I tried to simply add more rows to this, but the poster didn't appear as I wanted, and didn't want to make major changes to other people's work.  For this same reason, I didn't want to go changing the main Seasons page but it would be nice if it could be made a bit shorter.  The main Poster could, I suppose, be shrinked as well as the Seasons items.

jimiatnymbl commented:
I see if I can get some time soon to grab your latest branch and test.

As for the pane covering up a bit of the text, I wouldn't worry about it for now. I personally prefer getting something working to the user and tweaks can come later.  Just my opinion.

One option to think about (not now though), is now that you have "* Show Videos" .... Could you also do that with showing details....then the pane could be hidden completely.  Again I'd leave it as is for now, just thinking out loud 🙂

debmint commented:
>     2. The actor details appear to only search for movies.  So if the actor is in both movies and TV shows, it only shows the movies.  And if the actor is only in TV shows then the faded pane is empty (I can also share a screenshot if needed).  You'll notice the Web Client shows both Movies and TV shows.

I cannot understand why your system is acting this way unless your copy is not up-to-date.  My earliest renditions   As I said in my previous reply to this, My library is tiny, but I still have some actors who are in both.  My TV show actors have their TV appearances showing up,  The same with movies, and I have a few that do have entries for both.

debmint commented:
> I'll see if I can get some time soon to grab your latest branch and test.

If you pulled anyway recently, I forgot to check formatting and found that somehow I lost an "end function" in source/ShowScene.brs for my CreatePerson... function.  This occurred during your Initial support for photo viewing.  In order to not have this bug in all commits between then and now, I did an interactive rebase at that commit, so my history has again been altered.  Maybe no one pulled while this error was in effect.
 
> One option to think about (not now though), is now that you have "* Show Videos" .... Could you also do that with showing details....then the pane could be hidden completely. Again I'd leave it as is for now, just thinking out loud slightly_smiling_face

I could, I suppose ("Show Videos" is now gone with the new interface).  I had thought something along these same lines, but somehow it just doesn't seem natural.  Too, I'd think you'd want to make this the same for all the extras, so it is an option.

jimdogx commented:
> I cannot understand why your system is acting this way unless your copy is not up-to-date. My earliest renditions As I said in my previous reply to this, My library is tiny, but I still have some actors who are in both. My TV show actors have their TV appearances showing up, The same with movies, and I have a few that do have entries for both.

It may not be your code.  I believe the server I'm connecting to is 10.8 beta... so it might be the server.  But I haven't had time to investigate.

> If you pulled anyway recently...

I haven't pulled yet and might not get a chance until tomorrow night.  Looking forward to seeing the latest version. 👍

debmint commented:
> As for the pane covering up a bit of the text, I wouldn't worry about it for now. I personally prefer getting something working to the user and tweaks can come later. Just my opinion.

I've moved toe pane home position down lower - to 1014.  It now is below the "Season x" text although there is still overlap between the pane and the focused-highlight border.  The Row text still shows so hopefully the user should be able to realize that something is there,

debmint commented:
>     getPersonVideos("Episode", results, { MaxWidth: 502, MaxHeight: 300 })
> ```
> 
> But this does:
> 
> ```
>     getPersonVideos("Episode, Series", results, { MaxWidth: 502, MaxHeight: 300 })
> ```
> 
> However, the TV Show posters are stretched.

debmint commented:
> For some reason, the TV Shows on the server I'm connecting to, are of type "Series". So this doesn't work on mine:

My system sends the complete series as type "Series".  When you call for TVShowDetails - listing the Seasons, the type is "Season", and individual episodes are "Episode"

> But this does:
> 
>     getPersonVideos("Episode, Series", results, { MaxWidth: 502, MaxHeight: 300 })

On my system, this produces a mixture of episodes AND series.  Somewhere, either the Jellyfin Web app, or Emby show a row for both Series and Episodes.  That could be done, of course, and would be useful for one of the stars of a series, but would there be much benefit to seeing a series where an actor appeared only once?  It shouldn't be a problem to add another row to the RowList if you wanted to see the Series.  Just add another call to the task and have it make the above call with "Series" as the first parameter.  BTW, do you think I should move the stuff I added to home/LoadItemsTask out into its own TASK - probably in extras directory?  

> However, the TV Show posters are stretched.

Yes, that is understandable.  The poster for a Series is in portrait mode  I pulled a series to see if there were more than one image and in the one I pulled, at least, there was only one image besides backdrops.  If you mix them, as in the above function call, you'd have to set the width for the individual item, based upon the type.

If you can and have time, keep investigating and see if you can find out more about what is going on with your system.  I found a program called "insomnia" which works similar to "postman" where you can send queries to the server and it will display the json response.  I found this quite helpful in seeing the structure of a response.

jimdogx commented:
> I found a program called "insomnia" which works similar to "postman" where you can send queries to the server and it will display the json response. I found this quite helpful in seeing the structure of a response.

Yep, we use Insomnia at work and it's what I use to debug Jellyfin stuff.  Great tool.  It's how I found that the User/{active_user}/Items call was returning "Movies" and "Series" for the person.

p.s. I noticed you closed the PR.  Was that by accident?

TwitchBronBron commented:
A few thoughts:
1. It might be worth making the background a solid color. Take a look at how the movie description bleeds through and intersects with the actor names. 
![image](https://user-images.githubusercontent.com/2544493/156646903-fd7f744d-73a8-490d-a699-a070e5b3e168.png)

2. Getting to the "show more" for cutoff person descriptions was not intuitive. I had to press "up", when I would expect to press "right". Then, when I pressed "up", the "Favorite" button unfocused, but it appeared that focus was completely lost. Here are a few options:
     ![image](https://user-images.githubusercontent.com/2544493/156649451-5c8c1727-4972-4b46-92d2-7dcff4afab82.png)

   - press "right" to unfocus "Favorite" button, then add a background color to the "Show More" text. 
   - press "right" to focus the entire text block, and add a border around it. 

3. In the "Show More" popup, it's not immediately obvious to me that the scrollbox is focused and the "close" button is not. So I was pressing "OK" to try and close, but that does nothing. Perhaps the "OK" button could close the popup when the text box is focused.
![image](https://user-images.githubusercontent.com/2544493/156650266-01b3e74e-ade7-429b-86e2-28c8f28033bb.png)

debmint commented:
Thank you for the input.  I will try to address your concerns

>     1. It might be worth making the background a solid color

Yes.  I see that.   I have not attempted to fix this yet. but I set the opacity to 0.9.  I thought it might be nice to see what's behind, but probably is a distraction.  I will set it to full to see if that eliminates it

>     2. Getting to the "show more" for cutoff person descriptions was not intuitive. I had to press "up", when I would expect to press "right".

You press "down" to go below the "Favorites", so, to me, it would seem natural to press "up" to go above.  For what it's worth, I took my cue from the way Emby does it, although I realize that Jellyfin is not trying to emulate them, but it seems like a practical way to do this.

> Then, when I pressed "up", the "Favorite" button unfocused, but it appeared that focus was completely lost.

When you press "up" the text becomes a little more enhanced.  Possibly not enough difference to be obvious, and I had thought about this and will look into it.

> Here are a few options:       
>        * press "right" to unfocus "Favorite" button, then add a background color to the "Show More" text.

I had thought making the "Show More" text stand out would be a good thing but had not looked into how to do so, if you even can do it with the components I used.  I will look into doing this.

As far as using the "right" button, it just doesn't seem correct.  For one thing, although not necessarily applicable to this scene, the "Favorite" button is not a button standing alone, but is a member of a ButtonGroup.  I did this to be able to utilize an existing calback to process toggling the Favorite status of the Person both locally and on the server.  It is actually the equivalent of the jellyfin-roku-defined component "ButtonHorizGroup" which contains the row of buttons in a MovieDetails or TVSeasonDetails, where a "right" or "left" button click moves you  horizontally through the buttons.  So it seems to me that the way I did this is more consistent with the rest of the app.

>     3. In the "Show More" popup, it's not immediately obvious to me that the scrollbox is focused and the "close" button is not. So I was pressing "OK" to try and close, but that does nothing. Perhaps the "OK" button could close the popup when the text box is focused.

I think I need to display the "close" button as a grayed-out button when bot focused.  This is a StandardMessageDialog and it simply uses this component's API.  The "Close" button gains focus after you scroll to the bottom and press "down" again (Fast Forward of course scrolls a full page at once).  When the "Close" button gains focus, "OK" then closes the dialog.  "Back" will close the dialog at any time.  Closing with "OK" does seem intuitive, but it will take adding to the key press logic.  I tried adding a prompt about pressing "back" in the title, as you seem to have done with hint about "OK", it just seemed to overshadow the overview text.  On quick scan of the docs, I didn't come across any way of setting the font size for the title in this particular dialog.

Thanks for the input.  I will try to implement some of your suggestions.

neilsb reviewed (approved):
Lot of work gone into this.  Looking really good.  👍

debmint commented:
> Lot of work gone into this. Looking really good. +1

And I got some helpful suggestions for fixes, which I definitely appreciate.  I did negelect to move my additions out of home/LoadItemsTask, or rename the variable I added.  perhaps someone can clean up any errors I may have made.
